rasa/core/policies/rule_policy.py

```diff
@@ -415,40 +423,62 @@ def _check_prediction(
         domain: Domain,
         interpreter: NaturalLanguageInterpreter,
         gold_action_name: Text,
+        collect_sources: bool,
     ) -> Optional[Text]:
 
         predicted_action_name = self._predict_next_action(tracker, domain, interpreter)
-        if not predicted_action_name or predicted_action_name == gold_action_name:
-            return None
-
         # RulePolicy will always predict active_loop first,
         # but inside loop unhappy path there might be another action
-        if predicted_action_name == tracker.active_loop_name:
+        if (
+            predicted_action_name != gold_action_name
+            and predicted_action_name == tracker.active_loop_name
+        ):
             rasa.core.test.emulate_loop_rejection(tracker)
             predicted_action_name = self._predict_next_action(
                 tracker, domain, interpreter
             )
-            if not predicted_action_name or predicted_action_name == gold_action_name:
-                return None
+
+        if collect_sources:
```

Comment:
Can't we move up lines 441 to 447 to the top of the method to skip the actual prediction part? Should be a faster.

Reply:
or making it an `if else` thin within `_run_prediction_on_trackers`?

Reply:
the actual prediction part is necessary, because `predict_action_probabilities` updates `self._prediction_source` that is used in new lookup

Reply:
and we cannot simply call `predict_action_probabilities` because we need to emulate form rejection